models/User.go

```diff
@@ -1023,13 +1067,95 @@ func PlaceBidOrder(userId uint32, bid *Bid) (uint32, error) {
 			StockQuantity: bid.StockQuantity,
 		})
 		transactionsStream.SendTransaction(orderFeeTransaction.ToProto())
+		transactionsStream.SendTransaction(placeOrderTransaction.ToProto())
 
 		l.Infof("Sent through the datastreams")
-	}(bid, orderFeeTransaction)
+	}(bid, orderFeeTransaction, placeOrderTransaction)
 
 	return bid.Id, nil
 }
 
+// SaveAskCancelOrderTransaction creates a CancelOrderTransaction for Ask orders and pushses to stream
+func SaveAskCancelOrderTransaction(askOrder *Ask, user *User, tx *gorm.DB) error {
+	var l = logger.WithFields(logrus.Fields{
+		"method":  "SaveAskCancelOrderTransaction",
+		"userId":  user.Id,
+		"orderId": askOrder.Id,
+	})
+
+	cancelOrderTransaction := GetTransactionRef(
+		user.Id,
+		askOrder.StockId,
+		CancelOrderTransaction,
+		int64(askOrder.StockQuantity-askOrder.StockQuantityFulfilled),
+		0,
+		0,
+	)
+
+	if err := tx.Save(cancelOrderTransaction).Error; err != nil {
+		tx.Rollback()
+		l.Errorf("Error while commiting %+v", err)
+		return err
+	}
+
+	go func(cancelOrderTransaction *Transaction) {
+		transactionsStream := datastreamsManager.GetTransactionsStream()
+		transactionsStream.SendTransaction(cancelOrderTransaction.ToProto())
+
+		l.Infof("Sent through the datastreams")
+	}(cancelOrderTransaction)
+
+	return nil
+}
+
+// SaveBidCancelOrderTransaction creates a CancelOrderTransaction for Ask orders and pushses to stream
+func SaveBidCancelOrderTransaction(bidOrder *Bid, user *User, tx *gorm.DB) error {
+	var l = logger.WithFields(logrus.Fields{
+		"method":  "SaveBidCancelOrderTransaction",
+		"userId":  user.Id,
+		"orderId": bidOrder.Id,
+	})
+
+	reservedCash, _, err := GetPlaceOrderTransactionDetails(bidOrder.Id, false)
+	if err != nil {
+		l.Errorf("Could not retrieve reserved cash. Error: %+v")
+		return err
+	}
+
+	stocksFulfilled := bidOrder.StockQuantity - bidOrder.StockQuantityFulfilled
+	reservedCash = int64(float64(reservedCash) * float64(stocksFulfilled) / float64(bidOrder.StockQuantity))
+	cancelOrderTransaction := GetTransactionRef(
+		user.Id,
+		bidOrder.StockId,
+		CancelOrderTransaction,
+		0,
+		0,
+		reservedCash,
+	)
+
+	user.Cash += uint64(reservedCash)
+	if err := tx.Save(user).Error; err != nil {
+		tx.Rollback()
```

Comment:
Don't rollback here. You should just return the error and the caller should handle the rollback.

Reply:
:+1: